File: src/insights.ts
/**
 * The one-line caption shown at reveal.
 *
 * The PRD gives AI an offline job here: "generate concise grounded
 * explanations". Grounded is the operative word, so every sentence below is
 * assembled from fields that are already in the database -- dates, regions,
 * object types, makers. Nothing is invented, nothing is fetched at play time,
 * and the caption never carries any weight in deciding the answer.
 *
 * Replacing this with an offline model pass means changing one function; the
 * rest of the codebase only ever reads the `insight` column.
 */

import * as dates from "./dates.ts";
import * as taxonomy from "./taxonomy.ts";

export interface InsightRow extends dates.DatedRow {
  title: string;
  artist?: string | null;
  medium?: string | null;
  classification?: string | null;
  region?: string | null;
  looks_modern?: number | boolean | null;
  year_mid?: number | null;
}

const form = (row: InsightRow): string =>
  taxonomy.formFor(row.medium, row.classification, row.title);

const place = (row: InsightRow): string | null => {
  const region = row.region;
  return !region || region === "Unknown" ? null : region;
};

const decade = (row: InsightRow): string => `${Math.floor((row.year_mid ?? 0) / 10) * 10}s`;

function surname(name: string | null | undefined): string | null {
  if (!name) return null;
  const cleaned = name.split("(")[0]!.trim().replace(/,+$/, "");
  return cleaned || null;
}

/**
 * The date to name in a sentence -- the museum's claim, not our midpoint.
 *
 * Same value the card prints, with the article a sentence needs: a card says
 * "6th century", a sentence says "the 6th century".
 */
function when(row: InsightRow): string {
  const text = dates.headline(row);
  return text.endsWith("century") || text.endsWith("century BC") ? `the ${text}` : text;
}

/** Build the reveal caption for one pair. */
export function forPair(
  earlier: InsightRow,
  later: InsightRow,
  gap: number,
  approximate: boolean,
): string {
  const gapText = dates.describeGap(gap, approximate);
  const earlyForm = form(earlier);
  const lateForm = form(later);
  const earlyPlace = place(earlier);
  const latePlace = place(later);
  const earlyYear = when(earlier);
  const lateYear = when(later);
  const earlyMaker = surname(earlier.artist);
  const lateMaker = surname(later.artist);

  // The genuinely surprising case: the older object is the one that looks new.
  if (earlier.looks_modern && !later.looks_modern) {
    return (
      `The ${earlyForm} is the older of the two — ${gapText}, ` +
      "even though it reads as the more modern object."
    );
  }

  if (gap <= 3) {
    const both = earlyForm === lateForm ? `two ${earlyForm}s` : `a ${earlyForm} and a ${lateForm}`;
    if (earlyPlace && latePlace && earlyPlace !== latePlace) {
      return `Near-contemporaries: ${both}, made in ${earlyPlace} and ${latePlace} ${gapText}.`;
    }
    return `Near-contemporaries: ${both}, ${gapText}.`;
  }

  if (earlyMaker && lateMaker && earlyMaker === lateMaker) {
    return `Both by ${earlyMaker}, ${gapText} — the ${earlyForm} came first.`;
  }

  if (gap >= 500) {
    return (
      `${dates.centuryLabel(earlier.year_mid ?? 0)} against ` +
      `${dates.centuryLabel(later.year_mid ?? 0)}: ${gapText}.`
    );
  }

  if (earlyPlace && latePlace && earlyPlace !== latePlace) {
    return `${earlyPlace}, ${earlyYear} against ${latePlace}, ${lateYear} — ${gapText}.`;
  }

  if (earlyForm === lateForm) {
    return `Two ${earlyForm}s ${gapText}: ${earlyYear}, then ${lateYear}.`;
  }

  if (
    dates.isExact(earlier) &&
    dates.isExact(later) &&
    Math.floor((earlier.year_mid ?? 0) / 10) === Math.floor((later.year_mid ?? 0) / 10)
  ) {
    return `Both ${decade(earlier)} work — the ${earlyForm} came first, ${gapText}.`;
  }

  if (earlyForm === "object" || lateForm === "object") {
    // Naming a shape we could not identify adds nothing; let the dates speak.
    const opening = earlyYear.charAt(0).toUpperCase() + earlyYear.slice(1);
    return `${opening}, then ${lateYear} — ${gapText}.`;
  }
  return `The ${earlyForm} of ${earlyYear}, then the ${lateForm} of ${lateYear} — ${gapText}.`;
}

File: pastperfect/insights.py
"""The one-line caption shown at reveal.

The PRD gives AI an offline job here: "generate concise grounded explanations".
Grounded is the operative word, so every sentence below is assembled from fields
that are already in the database -- dates, regions, object types, makers. Nothing
is invented, nothing is fetched at play time, and the caption never carries any
weight in deciding the answer.

Replacing this with an offline model pass means changing one function; the rest
of the codebase only ever reads the ``insight`` column.
"""

from __future__ import annotations

from . import dates, taxonomy


def _form(row: dict) -> str:
    return taxonomy.form_for(row.get("medium"), row.get("classification"), row.get("title"))


def _place(row: dict) -> str | None:
    region = row.get("region")
    return None if not region or region == "Unknown" else region


def _decade(row: dict) -> str:
    return f"{row['year_mid'] // 10 * 10}s"


def _surname(name: str | None) -> str | None:
    if not name:
        return None
    cleaned = name.split("(")[0].strip().rstrip(",")
    return cleaned or None


def for_pair(earlier: dict, later: dict, gap: int, approximate: bool) -> str:
    """Build the reveal caption for one pair."""
    gap_text = dates.describe_gap(gap, approximate)
    early_form, late_form = _form(earlier), _form(later)
    early_place, late_place = _place(earlier), _place(later)
    early_year = dates.format_year(earlier["year_mid"])
    late_year = dates.format_year(later["year_mid"])
    early_maker = _surname(earlier.get("artist"))
    late_maker = _surname(later.get("artist"))

    # The genuinely surprising case: the older object is the one that looks new.
    if earlier.get("looks_modern") and not later.get("looks_modern"):
        return (
            f"The {early_form} is the older of the two — {gap_text}, "
            f"even though it reads as the more modern object."
        )

    if gap <= 3:
        both = (
            f"two {early_form}s" if early_form == late_form
            else f"a {early_form} and a {late_form}"
        )
        if early_place and late_place and early_place != late_place:
            return f"Near-contemporaries: {both}, made in {early_place} and {late_place} {gap_text}."
        return f"Near-contemporaries: {both}, {gap_text}."

    if early_maker and late_maker and early_maker == late_maker:
        return f"Both by {early_maker}, {gap_text} — the {early_form} came first."

    if gap >= 500:
        return (
            f"{dates.century_label(earlier['year_mid'])} against "
            f"{dates.century_label(later['year_mid'])}: {gap_text}."
        )

    if early_place and late_place and early_place != late_place:
        return (
            f"{early_place}, {early_year} against {late_place}, {late_year} — {gap_text}."
        )

    if early_form == late_form:
        return f"Two {early_form}s {gap_text}: {early_year}, then {late_year}."

    if earlier["year_mid"] // 10 == later["year_mid"] // 10:
        return f"Both {_decade(earlier)} work — the {early_form} came first, {gap_text}."

    return f"The {early_form} of {early_year}, then the {late_form} of {late_year} — {gap_text}."
